// mindctl/src/ui/ascii.js
// ═══════════════════════════════════════════════════
// 🎨 mindctl — ASCII Art Library
// Handcrafted terminal art for every feature
// ═══════════════════════════════════════════════════

import chalk from 'chalk';
import gradient from 'gradient-string';

// ─── Main Logos ───────────────────────────────────

export const LOGOS = {
    full: `
  ███╗   ███╗██╗███╗   ██╗██████╗  ██████╗████████╗██╗
  ████╗ ████║██║████╗  ██║██╔══██╗██╔════╝╚══██╔══╝██║
  ██╔████╔██║██║██╔██╗ ██║██║  ██║██║        ██║   ██║
  ██║╚██╔╝██║██║██║╚██╗██║██║  ██║██║        ██║   ██║
  ██║ ╚═╝ ██║██║██║ ╚████║██████╔╝╚██████╗   ██║   ███████╗
  ╚═╝     ╚═╝╚═╝╚═╝  ╚═══╝╚═════╝  ╚═════╝   ╚═╝   ╚══════╝`,

    mini: `
  ╔╦╗╦╔╗╔╔╦╗╔═╗╔╦╗╦
  ║║║║║║║ ║║║   ║ ║
  ╩ ╩╩╝╚╝═╩╝╚═╝ ╩ ╩═╝`,

    brain: `
        ╭───────────╮
      ╭─┤  ◉     ◉  ├─╮
    ╭─┤  │  ╭─────╮  │  ├─╮
    │ │  │  │ ◕ ◕ │  │  │ │
    │ │  │  ╰──┬──╯  │  │ │
    ╰─┤  ╰────┤├────╯  ├─╯
      ╰───────┤├───────╯
              ╰╯
       YOUR BRAIN, OBSERVED`,
};

// ─── Section Headers ──────────────────────────────

export const HEADERS = {
    checkin: `
  ╔══════════════════════════════════╗
  ║  🎭  ░█▀▄▀█░█▀█░█▀█░█▀▄       ║
  ║      ░█░▀░█░█░█░█░█░█░█       ║
  ║      ░▀░░░▀░▀▀▀░▀▀▀░▀▀░       ║
  ║      CHECK-IN                   ║
  ╚══════════════════════════════════╝`,

    breathe: `
  ┌──────────────────────────────────┐
  │                                  │
  │    ╭─╮   ╭─╮   ╭─╮   ╭─╮      │
  │   ╱   ╲ ╱   ╲ ╱   ╲ ╱   ╲     │
  │  ╱     ╳     ╳     ╳     ╲    │
  │ ╱     ╱ ╲   ╱ ╲   ╱ ╲     ╲   │
  │╱     ╱   ╰─╯   ╰─╯   ╲     ╲  │
  │  🫁  B R E A T H E            │
  └──────────────────────────────────┘`,

    journal: `
  ┌──────────────────────────────────┐
  │  ┌────────────────────────────┐  │
  │  │ ═══════════════════════    │  │
  │  │ ═══════════════            │  │
  │  │ ════════════════════       │  │
  │  │ ══════════════             │  │
  │  │      📝                    │  │
  │  └────────────────────────────┘  │
  │    J O U R N A L                 │
  └──────────────────────────────────┘`,

    think: `
  ╭──────────────────────────────────╮
  │        ╭──────╮                  │
  │       ╱ ◉   ◉ ╲                 │
  │      │  ╭───╮  │   💭           │
  │      │  ╰───╯  │  ╱             │
  │       ╲  ───  ╱  ╱              │
  │        ╰──────╯ ╱               │
  │  🧠  T H I N K                  │
  ╰──────────────────────────────────╯`,

    talk: `
  ╭──────────────────────────────────╮
  │    ╭──────────╮                  │
  │    │ Hi! I'm  │                 │
  │    │ here for ├─╮               │
  │    │ you. 💙  │  ╲              │
  │    ╰──────────╯   ◉             │
  │                                  │
  │  💬  T A L K                     │
  ╰──────────────────────────────────╯`,

    burnout: `
  ╭──────────────────────────────────╮
  │         ╱╲                       │
  │        ╱  ╲   ╱╲                 │
  │    ╱╲ ╱    ╲ ╱  ╲    ╱╲         │
  │   ╱  ╳      ╳    ╲  ╱  ╲        │
  │  ╱  ╱ ╲    ╱ ╲    ╲╱    ╲       │
  │ ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓     │
  │  🔥  B U R N O U T              │
  ╰──────────────────────────────────╯`,

    stats: `
  ╭──────────────────────────────────╮
  │                     ╭──╮         │
  │              ╭──╮   │██│         │
  │        ╭──╮  │██│   │██│         │
  │  ╭──╮  │██│  │██│   │██│         │
  │  │██│  │██│  │██│   │██│         │
  │  ╰──╯  ╰──╯  ╰──╯   ╰──╯        │
  │  📊  S T A T S                   │
  ╰──────────────────────────────────╯`,

    meditate: `
  ╭──────────────────────────────────╮
  │            ✦                     │
  │         ✦  ◆  ✦                  │
  │        ✦ ◆ ★ ◆ ✦                │
  │         ✦  ◆  ✦                  │
  │            ✦                     │
  │                                  │
  │  🧘  M E D I T A T E            │
  ╰──────────────────────────────────╯`,

    crisis: `
  ╔══════════════════════════════════╗
  ║  ██╗ ██╗███████╗██╗     ██████╗ ║
  ║  ██║ ██║██╔════╝██║     ██╔══██╗║
  ║  ███████║█████╗  ██║     ██████╔╝║
  ║  ██╔══██║██╔══╝  ██║     ██╔═══╝ ║
  ║  ██║  ██║███████╗███████╗██║     ║
  ║  ╚═╝  ╚═╝╚══════╝╚══════╝╚═╝    ║
  ║  🚨  You're not alone.           ║
  ╚══════════════════════════════════╝`,

    sleep: `
  ╭──────────────────────────────────╮
  │           z                      │
  │          z  ☾                    │
  │         z                        │
  │     ╭────────╮                   │
  │     │ -  .  -│  ★  ·  ★         │
  │    ╱╰────────╯╲    ·            │
  │  ╱══════════════╲               │
  │  💤  S L E E P                   │
  ╰──────────────────────────────────╯`,

    habit: `
  ╭──────────────────────────────────╮
  │  ☐ → ☐ → ☑ → ☑ → ☑ → 🔥       │
  │                                  │
  │  Day 1  2   3   4   5   STREAK! │
  │                                  │
  │  🏆  H A B I T S                 │
  ╰──────────────────────────────────╯`,

    water: `
  ╭──────────────────────────────────╮
  │       ╭────╮                     │
  │       │~~~~│                     │
  │       │    │                     │
  │       │    │                     │
  │       ╰────╯                     │
  │  💧  W A T E R                   │
  ╰──────────────────────────────────╯`,

    daemon: `
  ╭──────────────────────────────────╮
  │      ╭──────────────╮            │
  │      │  👁        👁 │            │
  │      │    ╭════╮    │            │
  │      │    ║ 🛡️ ║    │            │
  │      │    ╰════╯    │            │
  │      ╰──────────────╯            │
  │  🛡️  G U A R D I A N             │
  ╰──────────────────────────────────╯`,

    config: `
  ╭──────────────────────────────────╮
  │        ╭──╮                      │
  │     ╭──┤  ├──╮                   │
  │  ╭──┤  │⚙ │  ├──╮               │
  │  │  ╰──┤  ├──╯  │               │
  │  ╰─────╰──╯─────╯               │
  │  ⚙️  C O N F I G                 │
  ╰──────────────────────────────────╯`,

    forecast: `
  ╭──────────────────────────────────╮
  │         ╲  │  ╱                  │
  │       ─── ☀ ───                 │
  │         ╱  │  ╲                  │
  │     ╭──────────╮                 │
  │    ╱  TOMORROW  ╲                │
  │   ╱══════════════╲               │
  │  🌤️  F O R E C A S T            │
  ╰──────────────────────────────────╯`,

    insights: `
  ╭──────────────────────────────────╮
  │      🔗─────────🔗               │
  │     ╱             ╲              │
  │    ●    r=0.87     ●             │
  │     ╲             ╱              │
  │      🔗─────────🔗               │
  │                                  │
  │  📊  I N S I G H T S             │
  ╰──────────────────────────────────╯`,

    report: `
  ╭──────────────────────────────────╮
  │  ┌──────────────────────────┐    │
  │  │  📊 WEEKLY REPORT        │    │
  │  │  ──────────────────      │    │
  │  │  ✅ Mood: +12%           │    │
  │  │  💤 Sleep: 7.2hrs        │    │
  │  │  🔥 Streak: 7 days       │    │
  │  └──────────────────────────┘    │
  │  📋  R E P O R T                 │
  ╰──────────────────────────────────╯`,

    focus: `
  ╭──────────────────────────────────╮
  │          (       )               │
  │           )     (                │
  │          (       )               │
  │       ╭─────────────╮            │
  │      ╱               ╲           │
  │     ╱                 ╲          │
  │    ╱═══════════════════╲         │
  │  🔥  Z E N  F O C U S            │
  ╰──────────────────────────────────╯`,

    flow: `
  ╭──────────────────────────────────╮
  │        ▒▒▒▒▒▒▒▒▒▒▒▒▒             │
  │      ▒▒            ▒▒            │
  │     ▒▒    ◒    ◒    ▒▒           │
  │     ▒▒      ◡      ▒▒            │
  │      ▒▒            ▒▒            │
  │        ▒▒▒▒▒▒▒▒▒▒▒▒▒             │
  │  🧬  F L O W  S T A T E          │
  ╰──────────────────────────────────╯`,

    quest: `
  ╭──────────────────────────────────╮
  │        / \                       │
  │       /   \        ✧             │
  │      /  ★  \                     │
  │     /_______\        ✧           │
  │    (_________)                   │
  │                                  │
  │  🏆  W E L L N E S S  Q U E S T  │
  ╰──────────────────────────────────╯`,

    review: `
  ╭──────────────────────────────────╮
  │        .───────────.             │
  │     .-'  (◉)   (◉)  '-.          │
  │    /   ─────────────   \         │
  │   |    (  🧠 SCAN  )    |        │
  │    \   ─────────────   /         │
  │     '-.             .-'          │
  │        '───────────'             │
  │  🤖  A I  M E T A - R E V I E W  │
  ╰──────────────────────────────────╯`,
};

// ─── Fireplace Animation Frames ───────────────────

export const FIREPLACE = [
    [
        '       (  )       ',
        '      (    )      ',
        '     (      )     ',
        '    (  🔥    )    ',
        '    ──────────    ',
        '   [          ]   ',
    ],
    [
        '      (    )      ',
        '     (  ( ) )     ',
        '    (  (🔥)  )    ',
        '     (      )     ',
        '    ──────────    ',
        '   [          ]   ',
    ],
    [
        '     (      )     ',
        '    (  (  )  )    ',
        '    ( ( 🔥 ) )    ',
        '     (      )     ',
        '    ──────────    ',
        '   [          ]   ',
    ],
];

// ─── Mood Face ASCII Art ──────────────────────────

export const MOOD_FACES = {
    amazing: [
        '  ╭───────╮  ',
        '  │ ★   ★ │  ',
        '  │  ╰─╯  │  ',
        '  │ ╰═══╯ │  ',
        '  ╰───────╯  ',
    ],
    happy: [
        '  ╭───────╮  ',
        '  │ ◕   ◕ │  ',
        '  │       │  ',
        '  │ ╰───╯ │  ',
        '  ╰───────╯  ',
    ],
    good: [
        '  ╭───────╮  ',
        '  │ ●   ● │  ',
        '  │       │  ',
        '  │ ╰──╯  │  ',
        '  ╰───────╯  ',
    ],
    okay: [
        '  ╭───────╮  ',
        '  │ ●   ● │  ',
        '  │       │  ',
        '  │ ────  │  ',
        '  ╰───────╯  ',
    ],
    meh: [
        '  ╭───────╮  ',
        '  │ •   • │  ',
        '  │       │  ',
        '  │ ════  │  ',
        '  ╰───────╯  ',
    ],
    low: [
        '  ╭───────╮  ',
        '  │ •   • │  ',
        '  │       │  ',
        '  │ ╭──╮  │  ',
        '  ╰───────╯  ',
    ],
    sad: [
        '  ╭───────╮  ',
        '  │ ·   · │  ',
        '  │   ·   │  ',
        '  │ ╭───╮ │  ',
        '  ╰───────╯  ',
    ],
    anxious: [
        '  ╭───────╮  ',
        '  │ ◉   ◉ │  ',
        '  │   ~   │  ',
        '  │ ╭───╮ │  ',
        '  ╰───────╯  ',
    ],
    stressed: [
        '  ╭───────╮  ',
        '  │ ×   × │  ',
        '  │   ·   │  ',
        '  │ ╭═══╮ │  ',
        '  ╰───────╯  ',
    ],
    angry: [
        '  ╭───────╮  ',
        '  │ ╲   ╱ │  ',
        '  │  ◉ ◉  │  ',
        '  │ ╰═══╯ │  ',
        '  ╰───────╯  ',
    ],
    peaceful: [
        '  ╭───────╮  ',
        '  │ ─   ─ │  ',
        '  │       │  ',
        '  │  ◡◡◡  │  ',
        '  ╰───────╯  ',
    ],
    grateful: [
        '  ╭───────╮  ',
        '  │ ♥   ♥ │  ',
        '  │       │  ',
        '  │ ╰═══╯ │  ',
        '  ╰───────╯  ',
    ],
    tired: [
        '  ╭───────╮  ',
        '  │ -   - │ z',
        '  │       │z ',
        '  │  ───  │  ',
        '  ╰───────╯  ',
    ],
    overwhelmed: [
        '  ╭───────╮  ',
        '  │ @   @ │  ',
        '  │ ~~~~~ │  ',
        '  │ ╭═══╮ │  ',
        '  ╰───────╯  ',
    ],
    hopeful: [
        '  ╭───────╮  ',
        '  │ ✦   ✦ │  ',
        '  │       │  ',
        '  │ ╰──╯  │  ',
        '  ╰───────╯  ',
    ],
};

// ─── Breathing Circles (Expanded) ─────────────────

export const BREATH_CIRCLES = {
    inhale: [
        [
            '            ●            ',
        ],
        [
            '          ╭───╮          ',
            '          │   │          ',
            '          ╰───╯          ',
        ],
        [
            '        ╭───────╮        ',
            '        │       │        ',
            '        │       │        ',
            '        ╰───────╯        ',
        ],
        [
            '      ╭───────────╮      ',
            '      │           │      ',
            '      │           │      ',
            '      │           │      ',
            '      ╰───────────╯      ',
        ],
        [
            '    ╭───────────────╮    ',
            '    │               │    ',
            '    │               │    ',
            '    │               │    ',
            '    │               │    ',
            '    ╰───────────────╯    ',
        ],
        [
            '  ╭───────────────────╮  ',
            '  │                   │  ',
            '  │                   │  ',
            '  │                   │  ',
            '  │                   │  ',
            '  │                   │  ',
            '  ╰───────────────────╯  ',
        ],
    ],
};

// ─── Achievement Badges ───────────────────────────

export const BADGES = {
    first_breath: `
  ╭─────╮
  │ 🫁  │  First Breath
  │  ★  │  Completed first breathing exercise
  ╰─────╯`,
    week_warrior: `
  ╭─────╮
  │ 🔥  │  Week Warrior
  │ ★★★ │  7-day wellness streak
  ╰─────╯`,
    thought_detective: `
  ╭─────╮
  │ 🧠  │  Thought Detective
  │  ★  │  First CBT thought record
  ╰─────╯`,
    zen_master: `
  ╭─────╮
  │ 🧘  │  Zen Master
  │★★★★★│  30 meditation sessions
  ╰─────╯`,
    vulnerability: `
  ╭─────╮
  │ 📝  │  Vulnerability
  │ ★★  │  20 journal entries
  ╰─────╯`,
    level_up: `
  ╭─────────────────╮
  │  ⚡ LEVEL UP! ⚡  │
  │                  │
  │   ╭──╮  ★  ★    │
  │   │▲▲│ ★ ★ ★    │
  │   │▼▼│  ★  ★    │
  │   ╰──╯          │
  ╰─────────────────╯`,
};

// ─── Wellness Icons (ASCII) ───────────────────────

export const WELLNESS_ART = {
    water_glass: [
        '  ╭────╮  ',
        '  │~~~~│  ',
        '  │≈≈≈≈│  ',
        '  │    │  ',
        '  ╰─┬┬─╯  ',
        '    ╰╯    ',
    ],
    eye: [
        '  ╭─────────╮  ',
        ' ╱  ╭─────╮  ╲ ',
        '│  ╱ ╭───╮ ╲  │',
        '│  │ │ ● │ │  │',
        '│  ╲ ╰───╯ ╱  │',
        ' ╲  ╰─────╯  ╱ ',
        '  ╰─────────╯  ',
    ],
    posture: [
        '      ○      ',
        '     /│\\     ',
        '    / │ \\    ',
        '      │      ',
        '     / \\     ',
        '    /   \\    ',
        '   ─     ─   ',
        '  GOOD POSTURE',
    ],
    stretch_neck: [
        '     ╭─╮      ',
        '    ╱   ╲     ',
        '   │  ○  │    ',
        '    ╲   ╱     ',
        '     │ │      ',
        '    ╱│ │╲     ',
        '   ╱ │ │ ╲    ',
        '     │ │      ',
        '    ╱   ╲     ',
    ],
    walk: [
        '      ○       ',
        '     ╱│╲      ',
        '    ╱ │ ╲     ',
        '      │       ',
        '     ╱ ╲      ',
        '    ╱   ╲     ',
        '   ╱     ╲    ',
        '  →  WALK  →  ',
    ],
    heart: [
        '   ╭──╮ ╭──╮   ',
        '  ╱    ╳    ╲  ',
        '  │    ♥    │  ',
        '  ╲        ╱  ',
        '   ╲      ╱   ',
        '    ╲    ╱    ',
        '     ╲  ╱     ',
        '      ╲╱      ',
    ],
    shield: [
        '   ╭──────╮   ',
        '  ╱ ╭────╮ ╲  ',
        '  │ │ 🛡️ │ │  ',
        '  │ │    │ │  ',
        '  │ ╰────╯ │  ',
        '  ╲        ╱  ',
        '   ╲      ╱   ',
        '    ╰────╯    ',
    ],
    moon: [
        '       ★        ',
        '    ╭────╮   ·  ',
        '   ╱      ╲    ',
        '  │   ☾    │ ★ ',
        '   ╲      ╱    ',
        '    ╰────╯  ·  ',
        '        ★      ',
    ],
    sun: [
        '    ╲ │ ╱     ',
        '   ──╭─╮──    ',
        '     │☀│      ',
        '   ──╰─╯──    ',
        '    ╱ │ ╲     ',
    ],
};

// ─── Companion Evolution Art (Enhanced) ───────────

export const COMPANION_ART = {
    egg: {
        happy: [
            '      ╭────╮      ',
            '     ╱  ◕◕  ╲     ',
            '    │   ‿‿   │    ',
            '    │        │    ',
            '     ╲      ╱     ',
            '      ╰────╯      ',
        ],
        sad: [
            '      ╭────╮      ',
            '     ╱  ·_·  ╲     ',
            '    │   ︵   │    ',
            '    │        │    ',
            '     ╲      ╱     ',
            '      ╰────╯      ',
        ],
        excited: [
            '      ╭────╮  ✦   ',
            '     ╱  ★★  ╲     ',
            '    │   ▽   │  ★  ',
            '    │        │    ',
            '     ╲      ╱     ',
            '      ╰────╯      ',
        ],
        neutral: [
            '      ╭────╮      ',
            '     ╱  ••  ╲     ',
            '    │   ──   │    ',
            '    │        │    ',
            '     ╲      ╱     ',
            '      ╰────╯      ',
        ],
    },
    kitten: {
        happy: [
            '     ╱╲   ╱╲      ',
            '    ╱  ╲_╱  ╲     ',
            '    │  ^.^  │     ',
            '    │  ╰◡╯  │     ',
            '     ╲     ╱      ',
            '    ╱│     │╲     ',
            '   ╱ │     │ ╲    ',
            '      ╰───╯       ',
        ],
        sad: [
            '     ╱╲   ╱╲      ',
            '    ╱  ╲_╱  ╲     ',
            '    │  ;_;  │     ',
            '    │  ╭─╮  │     ',
            '     ╲     ╱      ',
            '    ╱│     │╲     ',
            '   ╱ │     │ ╲    ',
            '      ╰───╯       ',
        ],
        excited: [
            '     ╱╲   ╱╲  ✦   ',
            '    ╱  ╲_╱  ╲     ',
            '    │  ◕‿◕  │  ♥  ',
            '    │  ╰▽╯  │     ',
            '     ╲  ♥  ╱      ',
            '    ╱│     │╲     ',
            '   ╱ │     │ ╲    ',
            '      ╰───╯       ',
        ],
        neutral: [
            '     ╱╲   ╱╲      ',
            '    ╱  ╲_╱  ╲     ',
            '    │  •.•  │     ',
            '    │  ───  │     ',
            '     ╲     ╱      ',
            '    ╱│     │╲     ',
            '   ╱ │     │ ╲    ',
            '      ╰───╯       ',
        ],
        worried: [
            '     ╱╲   ╱╲      ',
            '    ╱  ╲_╱  ╲     ',
            '    │  ó.ò  │     ',
            '    │  ╭∼╮  │     ',
            '     ╲     ╱      ',
            '    ╱│     │╲     ',
            '   ╱ │     │ ╲    ',
            '      ╰───╯       ',
        ],
    },
    cat: {
        happy: [
            '      ╱╲     ╱╲      ',
            '     ╱  ╲   ╱  ╲     ',
            '    ╱    ╲─╱    ╲    ',
            '    │   ^   ^   │    ',
            '    │    ╰◡╯    │    ',
            '    │   ╰═══╯   │    ',
            '     ╲   ♥ ♥   ╱     ',
            '    ╭─╰───────╯─╮    ',
            '    │  ╱     ╲  │    ',
            '    ╰─╯       ╰─╯    ',
        ],
        excited: [
            '      ╱╲     ╱╲  ★   ',
            '     ╱  ╲   ╱  ╲     ',
            '    ╱    ╲─╱    ╲ ✦  ',
            '    │   ★   ★   │    ',
            '    │    ╰▽╯    │    ',
            '    │   ╰═══╯   │    ',
            '     ╲   ♥ ♥   ╱     ',
            '    ╭─╰───────╯─╮ ✦  ',
            '    │  ╱     ╲  │    ',
            '    ╰─╯       ╰─╯    ',
        ],
        sad: [
            '      ╱╲     ╱╲      ',
            '     ╱  ╲   ╱  ╲     ',
            '    ╱    ╲─╱    ╲    ',
            '    │   ·   ·   │    ',
            '    │    ╭─╮    │    ',
            '    │   ╭═══╮   │    ',
            '     ╲         ╱     ',
            '    ╭─╰───────╯─╮    ',
            '    │  ╱     ╲  │    ',
            '    ╰─╯       ╰─╯    ',
        ],
        neutral: [
            '      ╱╲     ╱╲      ',
            '     ╱  ╲   ╱  ╲     ',
            '    ╱    ╲─╱    ╲    ',
            '    │   •   •   │    ',
            '    │    ───    │    ',
            '    │           │    ',
            '     ╲         ╱     ',
            '    ╭─╰───────╯─╮    ',
            '    │  ╱     ╲  │    ',
            '    ╰─╯       ╰─╯    ',
        ],
    },
    lion: {
        happy: [
            '    ╔═══╦═══╦═══╗       ',
            '   ╔╝   ╱╲   ╱╲ ╚╗      ',
            '   ║   ╱  ╲─╱  ╲ ║      ',
            '   ║   │  ^ω^  │ ║      ',
            '   ╚╗  │  ╰◡╯  │╔╝      ',
            '    ╚══╰────────╯══╗     ',
            '        │ ♥ ♥ │          ',
            '       ╱│     │╲        ',
            '      ╱ ╰─────╯ ╲       ',
        ],
        excited: [
            '  ★ ╔═══╦═══╦═══╗ ★     ',
            '   ╔╝   ╱╲   ╱╲ ╚╗  ✦   ',
            '   ║   ╱  ╲─╱  ╲ ║      ',
            '   ║   │  ★▽★  │ ║      ',
            '   ╚╗  │  ╰◡╯  │╔╝   ★  ',
            '    ╚══╰────────╯══╗     ',
            '        │ ✦ ✦ │          ',
            '       ╱│     │╲        ',
            '      ╱ ╰─────╯ ╲       ',
        ],
        neutral: [
            '    ╔═══╦═══╦═══╗       ',
            '   ╔╝   ╱╲   ╱╲ ╚╗      ',
            '   ║   ╱  ╲─╱  ╲ ║      ',
            '   ║   │  •_•  │ ║      ',
            '   ╚╗  │  ───  │╔╝      ',
            '    ╚══╰────────╯══╗     ',
            '        │     │          ',
            '       ╱│     │╲        ',
            '      ╱ ╰─────╯ ╲       ',
        ],
    },
    dragon: {
        happy: [
            '              ╱╲          ',
            '          ╱╲ ╱  ╲ ╱╲      ',
            '         ╱  ╳    ╳  ╲     ',
            '        ╱  ╱╲  ╱╲  ╲    ',
            '       │  │ ◕  ◕ │  │    ',
            '       │  │  ╰▽╯  │  │    ',
            '      ╱╲  │ ♥♥♥♥  │ ╱╲   ',
            '     ╱  ╲─╰────────╯╱  ╲  ',
            '    ╱  ╱ ═══════════ ╲  ╲ ',
            '   ╱  ╱  │ ╱╲  ╱╲ │  ╲  ╲',
            '  ┘  ┘   ╰╱  ╲╱  ╲╯   └  └',
        ],
        excited: [
            '    ★         ╱╲      ★   ',
            '  ✦       ╱╲ ╱  ╲ ╱╲  ✦  ',
            '         ╱  ╳    ╳  ╲     ',
            '   ★    ╱  ╱╲  ╱╲  ╲  ★ ',
            '       │  │ ★  ★ │  │    ',
            '       │  │  ╰▽╯  │  │    ',
            '  ✦  ╱╲  │ ★★★★  │ ╱╲ ✦ ',
            '     ╱  ╲─╰────────╯╱  ╲  ',
            '    ╱  ╱ ═══════════ ╲  ╲ ',
            '   ╱  ╱  │ ╱╲  ╱╲ │  ╲  ╲',
            '  ┘  ┘   ╰╱  ╲╱  ╲╯   └  └',
        ],
        neutral: [
            '              ╱╲          ',
            '          ╱╲ ╱  ╲ ╱╲      ',
            '         ╱  ╳    ╳  ╲     ',
            '        ╱  ╱╲  ╱╲  ╲    ',
            '       │  │ •  • │  │    ',
            '       │  │  ───  │  │    ',
            '      ╱╲  │       │ ╱╲   ',
            '     ╱  ╲─╰────────╯╱  ╲  ',
            '    ╱  ╱ ═══════════ ╲  ╲ ',
            '   ╱  ╱  │ ╱╲  ╱╲ │  ╲  ╲',
            '  ┘  ┘   ╰╱  ╲╱  ╲╯   └  └',
        ],
    },
};

// ─── Decorative Frames ────────────────────────────

export const FRAMES = {
    simple: (text) => {
        const lines = text.split('\n');
        const maxLen = Math.max(...lines.map(l => l.length));
        const top = '  ╭' + '─'.repeat(maxLen + 2) + '╮';
        const bot = '  ╰' + '─'.repeat(maxLen + 2) + '╯';
        const body = lines.map(l => '  │ ' + l.padEnd(maxLen) + ' │');
        return [top, ...body, bot].join('\n');
    },
    double: (text) => {
        const lines = text.split('\n');
        const maxLen = Math.max(...lines.map(l => l.length));
        const top = '  ╔' + '═'.repeat(maxLen + 2) + '╗';
        const bot = '  ╚' + '═'.repeat(maxLen + 2) + '╝';
        const body = lines.map(l => '  ║ ' + l.padEnd(maxLen) + ' ║');
        return [top, ...body, bot].join('\n');
    },
    stars: (text) => {
        const lines = text.split('\n');
        const maxLen = Math.max(...lines.map(l => l.length));
        const border = '  ✦' + '═'.repeat(maxLen + 2) + '✦';
        const body = lines.map(l => '  ║ ' + l.padEnd(maxLen) + ' ║');
        return [border, ...body, border].join('\n');
    },
};

// ─── Utility: Render with Color ───────────────────

export function renderAscii(art, color = '#4ECDC4') {
    if (Array.isArray(art)) {
        return art.map(line => chalk.hex(color)(line)).join('\n');
    }
    return chalk.hex(color)(art);
}

export function renderGradientAscii(art, colors = ['#6C63FF', '#E040FB', '#FF4081']) {
    const grad = gradient(colors);
    if (Array.isArray(art)) {
        return art.map(line => grad(line)).join('\n');
    }
    return grad(art);
}

// ─── Utility: Show Header ─────────────────────────

export function showHeader(name, color = '#4ECDC4') {
    const header = HEADERS[name];
    if (header) {
        console.log(renderAscii(header, color));
    }
}

// ─── Utility: Show Mood Face ──────────────────────

export function showMoodFace(mood, color) {
    const face = MOOD_FACES[mood] || MOOD_FACES.okay;
    const c = color || '#4ECDC4';
    console.log(renderAscii(face, c));
}

// ─── Dividers & Decorators ────────────────────────

export const DIVIDERS = {
    wave: '  ∼∼∼∼∼∼∼∼∼∼∼∼∼∼∼∼∼∼∼∼∼∼∼∼∼∼∼∼∼∼∼∼∼∼∼∼∼∼∼∼',
    dots: '  · · · · · · · · · · · · · · · · · · · · ·',
    stars: '  ✦ · ✦ · ✦ · ✦ · ✦ · ✦ · ✦ · ✦ · ✦ · ✦ ·',
    dash: '  ┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄',
    thick: '  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━',
    double: '  ════════════════════════════════════════',
    zen: '  ─ · ─ · ─ · ─ · ─ · ─ · ─ · ─ · ─ · ─',
    fancy: '  ╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌',
};

// ─── Success / Completion Art ─────────────────────

export const COMPLETION = {
    checkmark: [
        '      ╱    ',
        '     ╱     ',
        '╲   ╱      ',
        ' ╲ ╱       ',
        '  V        ',
    ],
    trophy: [
        '     ╭───╮      ',
        '  ╭──┤ ★ ├──╮   ',
        '  │  ╰───╯  │   ',
        '  ╰────┬────╯   ',
        '       │        ',
        '    ╭──┴──╮     ',
        '    ╰─────╯     ',
    ],
    star_burst: [
        '       ★       ',
        '    ★ ╱│╲ ★    ',
        '    ╱  │  ╲    ',
        ' ★─── ★ ───★  ',
        '    ╲  │  ╱    ',
        '    ★ ╲│╱ ★    ',
        '       ★       ',
    ],
    fireworks: [
        '  ✦    ★    ✦  ',
        '    ·  ·  ·    ',
        ' ✦  ╲ │ ╱  ✦  ',
        '  ───╳★╳───    ',
        ' ✦  ╱ │ ╲  ✦  ',
        '    ·  ·  ·    ',
        '  ✦    ★    ✦  ',
    ],
};

export default {
    LOGOS, HEADERS, MOOD_FACES, BREATH_CIRCLES, BADGES,
    WELLNESS_ART, COMPANION_ART, FRAMES, DIVIDERS, COMPLETION,
    renderAscii, renderGradientAscii, showHeader, showMoodFace,
};
